nat/traversal/pinger.go

```diff
@@ -312,36 +269,66 @@ func (p *Pinger) pingTargetConsumer(pingParams *Params) {
 		return
 	}
 
-	log.Info().Msgf("Ping target received: IP: %v, port: %v", pingParams.ConsumerPublicIP, pingParams.ConsumerPort)
-	if !p.natProxy.isAvailable(pingParams.ProxyPortMappingKey) {
-		log.Warn().Msgf("NATProxy is not available for this transport protocol key %v", pingParams.ProxyPortMappingKey)
-		return
-	}
+	stop := make(chan struct{})
+	defer close(stop)
 
-	conn, err := p.getConnection(pingParams.ConsumerPublicIP, pingParams.ConsumerPort, pingParams.ProviderPort)
+	conn, err := p.multiPing(pingParams.IP, pingParams.ProviderPorts, pingParams.ConsumerPorts, 2, stop)
 	if err != nil {
-		log.Error().Err(err).Msg("Failed to get connection")
+		log.Err(err).Msg("Failed to ping remote peer")
 		return
 	}
 
-	pingStop := make(chan struct{})
-	defer close(pingStop)
-	go func() {
-		err := p.ping(conn, pingStop)
-		if err != nil {
-			log.Warn().Err(err).Msg("Error while pinging")
-		}
-	}()
-
-	err = p.pingReceiver(conn, pingParams.Cancel)
+	err = ipv4.NewConn(conn).SetTTL(128)
 	if err != nil {
-		log.Error().Err(err).Msg("Ping receiver error")
+		log.Error().Err(err).Msg("Failed to set connection TTL")
 		return
 	}
 
-	p.eventPublisher.Publish(event.AppTopicTraversal, event.BuildSuccessfulEvent(StageName))
+	conn.Write([]byte("Using this connection"))
 
+	p.eventPublisher.Publish(event.AppTopicTraversal, event.BuildSuccessfulEvent(StageName))
 	log.Info().Msg("Ping received, waiting for a new connection")
 
 	go p.natProxy.handOff(pingParams.ProxyPortMappingKey, conn)
 }
+
+func (p *Pinger) multiPing(remoteIP string, localPorts, remotePorts []int, initialTTL int, stop <-chan struct{}) (*net.UDPConn, error) {
+	if len(localPorts) != len(remotePorts) {
+		return nil, errors.New("number of local and remote ports does not match")
+	}
+
+	type res struct {
+		conn *net.UDPConn
+		err  error
+	}
+
+	ch := make(chan res, len(localPorts))
+
+	for i := range localPorts {
+		go func(i int) {
+			conn, err := p.singlePing(remoteIP, localPorts[i], remotePorts[i], initialTTL+i, stop)
+			ch <- res{conn, err}
+		}(i)
+	}
+
+	// First responce wins. Other are not important.
```

Comment:
First might be connection establishment failure, then others might be important.

Reply:
The real error could happen only if we failed to create connection which is significant.
In other cases, this will be just a timeout error, which means that whole pinger timed out.